doc/smart-switch/high-availability/smart-switch-ha-hld.md

```diff
@@ -0,0 +1,2208 @@
+
+Since data plane channel is inline, it is very important to make sure whenever gray network failure happens (traffic being blackholed partially), we should avoid causing the traffic for the entire ENI to be dropped.
+
+To provide HA for DPU-to-DPU communication, we can also leverage our ECMP network setup. When sending the tunneled packet, we must calculate the source port from the 5 tuples of the inner packet and fold it into a port range specified by SONiC. This helps us to spread the packets across all possible paths and avoid the blackhole problem.
+
+###### 4.2.3.0.3. (Optional) Multi-path data plane availability tracking
+
+Multi-path communication is great, but it also introduces gray failures. For example, flow replication fails 10% of the time instead of 100% of time. 
+
+To help us further avoid link problems, DPUs could track the failures of each possible path and avoid using it when it is possible. This can be done by using actively probing (see "[DPU-to-DPU liveness probe](#73-dpu-to-dpu-liveness-probe)") or tracking the reachability of each data path by checking the response of data plane channel requests, e.g. flow replication.
+
+And here is an example of possible implementation based-on source port rotation that can be considered:
+
+1. Since data plane channel packet has VxLan encap as outermost header, and network routes the packet by outermost header too, the data path entropy is determined only by the source port of the outermost packet (protocol, destination IP/port, source IP are all determined already).
+2. To leverage ECMP setup, we can maintain a pool of source ports to use, and each flow can pick up a random port from the pool. The algorithm for port selection doesn’t matter. As long as we can spread the packet across all ports, we should be good.
+3. The flow replication packet will be sent out with the selected source port in outermost packet, as well as in the metadata. And on the other side, the ack packet can use the same source port too. This binds the packets for this flow to a single data path.
+4. If one of the links went down, all the packets with certain source ports will start to fail. Hence, if we could track the failure for each port in the pool, we can avoid using it.
+5. Alternatively, the DPU-to-DPU probe can use the same ports in the pool, in order to reflect the data path problem. This also helps us to see when the data path is recovered, and we can resume using this port again.
+
+#### 4.2.4. VM to DPU data plane
+
+Again, it is very important to avoid completely killing the VM’s connectivity if one link is blackholing traffic. Hence, when VM sends traffic to DPU, we should also make it leverage ECMP setup.
+
+When VM sends packet out, the packet will be encap’ed by VxLan (see below). Hence, similar to DPU-to-DPU data plane channel, we can also set the VxLan source port by calculating the hash of the inner packet. Then, different flow will go through different data path to reach the SmartSwitch.
```

Comment:
See above comment about source ports affecting HASH-based ECMP egress port selection.

Reply:
Have updated the doc now to make the algorithm align.